utils/yoke/src/is_covariant.rs

```diff
@@ -0,0 +1,133 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[cfg(feature = "alloc")]
+use alloc::{
+    borrow::{Cow, ToOwned},
+    boxed::Box,
+    rc::Rc,
+};
+
+/// A type implementing `IsCovariant<'a>` is covariant with respect to lifetime `'a`.
+///
+/// All types on which it is safe to implement [`Yokeable`] can also safely
+/// implement `IsCovariant`. The difference is that `Yokeable` is defined on the `'static` version
+/// of the type, whereas `IsCovariant` is defined on the `'a` version of the type. This makes it
+/// useful in trait bounds when a type needs to be covariant for another unsafe operation.
+///
+/// The primary use case is to safely perform lifetime casting on trait objects (`dyn Trait`).
+/// This enables a type-erased [`Yoke`] consisting of only trait objects.
+///
+/// # Implementation safety
+///
```

Comment:
This type is safe to implement when the type has only covariant lifetime parameters which are all covered by `'a`.

(Crucial distinction: it is _not_ safe for `impl IsCovariant<'a> for Foo<'a, 'b>`, I think)

Reply:
Discussed